packages/xo-web/src/xo-app/hub/resource.js

```diff
@@ -0,0 +1,257 @@
+import _ from 'intl'
+import ActionButton from 'action-button'
+import decorate from 'apply-decorators'
+import Icon from 'icon'
+import React from 'react'
+import { Card, CardBlock, CardHeader } from 'card'
+import { Col, Row } from 'grid'
+import { alert, form } from 'modal'
+import { connectStore, formatSize, getXoaPlan } from 'utils'
+import { createGetObjectsOfType } from 'selectors'
+import { downloadAndInstallResource, deleteTemplates } from 'xo'
+import { error, success } from 'notification'
+import { find, filter } from 'lodash'
+import { injectState, provideState } from 'reaclette'
+import { withRouter } from 'react-router'
+
+import ResourceForm from './resource-form'
+
+const subscribeAlert = () =>
+  alert(
+    _('hubResourceAlert'),
+    <div>
+      <p>
+        {_('considerSubscribe', {
+          link: 'https://xen-orchestra.com',
+        })}
+      </p>
+    </div>
+  )
+
+export default decorate([
+  withRouter,
+  connectStore(() => {
+    const getTemplates = createGetObjectsOfType('VM-template').sort()
+    const getPools = createGetObjectsOfType('pool')
+    return {
+      templates: getTemplates,
+      pools: getPools,
+      hubInstallLoadingState: state => state.hubInstallLoadingState,
+    }
+  }),
+  provideState({
+    initialState: () => ({
+      selectedInstallPools: [],
+    }),
+    effects: {
+      async install() {
+        const {
+          id,
+          name,
+          namespace,
+          setHubInstallLoadingState,
+          version,
+        } = this.props
+        const {
+          hubInstallLoadingState,
+          isTemplateInstalled,
+          isFromSources,
+        } = this.state
+        if (isFromSources) {
+          subscribeAlert()
+          return
+        }
+        const resourceParams = await form({
+          render: props => (
+            <ResourceForm
+              install
+              multi
+              poolPredicate={isTemplateInstalled}
+              {...props}
+            />
+          ),
+          header: (
+            <span>
+              <Icon icon='add-vm' /> {name}
+            </span>
+          ),
+          size: 'medium',
+        })
+
+        setHubInstallLoadingState({
+          ...hubInstallLoadingState,
+          [id]: true,
+        })
+        for (const pool of resourceParams.pools) {
+          try {
+            await downloadAndInstallResource({
+              namespace,
+              id,
+              version,
+              sr: pool.default_SR,
+            })
```

Comment:
I'm not sure [your comment](https://github.com/vatesfr/xen-orchestra/pull/4442#discussion_r327638395) is still pointing to the correct line and for some obscure GitHub reason, I can't reply to it but anyway: I was talking about the `downloadAndInstallResource` calls. Right now, you're calling them sequentially (ie waiting for one to finish before starting the next one) but if there's nothing preventing you from calling them in parallel, you should do it:
```js
await Promise.all(pools.map(
  pool => downloadAndInstallResource(...)
))
```